### ISSUE TITLE
NotSupportedException When Joining Projections

### ISSUE THREAD

**Issue by claudiobernasconi** [nhibernate/nhibernate-core#1601]
This issue was originally posted on [StackOverflow ](https://stackoverflow.com/questions/48992552/nhibernate-notsupportedexception-when-joining-projections)with more details including a runnable test project. I did not get any community feedback within two weeks and therefore thinkt that this actually is a bug and try to report it here for clarification.

We experience an unexpected behavior of the NHibernate LINQ provider. We try to join two entities and select a projection each which sometimes results in a NotSupportedException.

[Detailed description of the issue and runnable code example](https://stackoverflow.com/questions/48992552/nhibernate-notsupportedexception-when-joining-projections).

_Please let me know if I should include the entire text in this GitHub issue as well or if you prefer the link the original post on StackOverflow._

Thank you for any feedback.

**Comment by fredericDelaporte** [comment]
I have not looked into the details on StackOverflow, but if you try joining two entities not by the mean of any mapped association, the error seems quite self explanatory.

Your report here currently looks more as a support request than as a bug/feature request: its place then belong to [NHUser groups](http://nhibernate.info/groups.html) (or StackOverflow, but you have already tried that).

**Comment by claudiobernasconi** [comment]
@fredericDelaporte Thank you very much for your feedback. I really appreciate your help. I am aware of the fact that this use case might not be a general use case.

In my opinion I cannot decide if this is a support request or a bug. If you read through the question posted on StackOverflow you can see that it works in one way but does not the other way. This seems strange to me and looks like a bug. Maybe there is a logical explanation why this behavior is as it is? 

Please take a look at the question in more detail if you want to solve our problem. I can assure you that we already put a lot of effort into it and are pretty familiar with nHiberate. We use it for more than 5 years in production.

**Comment by fredericDelaporte** [comment]
So some forms of join without using an association may work with Linq-2-NH, I was not knowing it. I have always considered this was to be done with QueryOver instead.

Determining if your case is a bug or a feature request would require to check what is going on in the provider code in both cases. (I personally do not intend to do it any time soon. Other contributors may do it.)

**Comment by hazzik** [comment]
Linq expressions are translated to HQL. HQL supports only joins by navigation properties and does not support SQL joins. HQL also supports select from multiple tables (cross join). So, inner join on unrelated tables can be emulated by cross join + where clause.

Linq translator does some simple _basic_ analysis to determine if the expression can be substituted by cross join + where clause, navigation join, etc. 

We would not be able to support all join types without extending the HQL grammar. 

So, each expression class should be considered in isolation. It's always not a bug, as NotSupportedException says what it is - this query is not supported. So, for every case it's either "could be done" - it's an improvement. Or "can not be done" - won't fix.

**Comment by claudiobernasconi** [comment]
@hazzik First of all thank you for giving me the detailed internals on how this works under the hood. The first query posted in the original question results in a cross join. According to your explanation, this makes sense to me. 

The question remains why the second query does not translate into a cross join as well? If you take a look at the queries, you will recognize that there is only a difference in the left and the right side of the cross join.

In my opinion, the second query should translate in a cross join as well if the first query successfully did. It would look something like this (order of the statements are reversed):

```
select player.Player_Id, team.Team_Id, team.Shortname, player.Name
    from DalTest_Player player, DalTest_Team team 
    where team.Team_Id = player.Team_Id
```

In other words: If it is not an intended behavior of the Linq provider I do not understand how it works in the first query but does not in the second?

However, I accept your answer that our intention is not supported and maybe never will be supported in the future.

**Comment by hazzik** [comment]
>The question remains why the second query does not translate into a cross join as well?

Analyzer is not intelligent enough to recognize this. You are more than welcome to improve it.

**Comment by hazzik** [comment]
@bahusoid does HQL support this now?

**Comment by bahusoid** [comment]
What hql do you have in mind? It looks like join on subquery to me:
```sql
 join (select  te.Id, te.ShortName TeamEntity te) team on team.Id = player.TeamId      
```
So I would say no. But maca88 is already [working on it](https://github.com/nhibernate/nhibernate-core/issues/1143#issuecomment-692822338)
